### PR TITLE
[CFM-1138] No JS fallback for Activity Stream.

### PR DESCRIPTION
**Pull request by ybaras**
[CFM-1138](https://issuetracker.amplexor.com/jira/browse/CFM-1138)

### CODE REVIEW

**Comment by ybaras** [comment]
No JS:

anim

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views_default.inc [hunk 1]
```diff
@@ -11,6 +11,133 @@ function c4m_message_views_default_views() {
...
+  $handler->display->display_options['filters']['field_group_node_target_id']['field'] = 'field_group_node_target_id';
+  $handler->display->display_options['filters']['field_group_node_target_id']['value']['value'] = '%1';
+  $handler->display->display_options['filters']['field_group_node_target_id']['group'] = 2;
+  /* Filter criterion: Field: Topics (c4m_vocab_topic) */
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['id'] = 'c4m_vocab_topic_tid';
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['table'] = 'field_data_c4m_vocab_topic';
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['field'] = 'c4m_vocab_topic_tid';
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['relationship'] = 'field_node_target_id';
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['value'] = array(
+    '327' => '327',
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views_default.inc]: this is probably wrong

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views_default.inc [hunk 2]
```diff
@@ -11,6 +11,133 @@ function c4m_message_views_default_views() {
   $export = array();
 
   $view = new view();
+  $view->name = 'global_activity_stream';
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views_default.inc]: add _js_fallback_ as suffix

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views_default.inc [hunk 3]
```diff
@@ -54,16 +181,6 @@ function c4m_message_views_default_views() {
   $handler->display->display_options['sorts']['timestamp']['table'] = 'message';
   $handler->display->display_options['sorts']['timestamp']['field'] = 'timestamp';
   $handler->display->display_options['sorts']['timestamp']['order'] = 'DESC';
-  /* Contextual filter: Message: Group node (field_group_node) */
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views_default.inc]: revert this one

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/filter/filter_homepage.js [hunk 4]
```diff
@@ -11,6 +11,8 @@
       var filter = radios.val();
       window.location.search = '?filter=' + filter;
     });
+
+    $('body').addClass('jsenabled');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/filter/filter_homepage.js]: `jsenabled` => `js-enabled`

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views_default.inc [hunk 5]
```diff
@@ -11,6 +11,133 @@ function c4m_message_views_default_views() {
...
+  );
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['group'] = 2;
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['reduce_duplicates'] = TRUE;
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['type'] = 'select';
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['vocabulary'] = 'c4m_vocab_topic';
+  $handler->display->display_options['filters']['c4m_vocab_topic_tid']['hierarchy'] = 1;
+
+  /* Display: Block */
+  $handler = $view->new_display('block', 'Block', 'block');
+  $handler->display->display_options['block_description'] = 'Global Activity Stream For NoJS';
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views_default.inc]: ...fallback  (somewhere)

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/filter/filter_homepage.js [hunk 6]
```diff
@@ -11,6 +11,8 @@
       var filter = radios.val();
       window.location.search = '?filter=' + filter;
     });
+
+    $('body').addClass('js-enabled');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/filter/filter_homepage.js]: Move this to a _theme_preprocess_page_

**Comment by ybaras** [comment]
**Group activity stream -> no-js**

anim

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/plugins/content_types/activity_stream/activity_stream.inc [hunk 7]
```diff
@@ -24,6 +24,7 @@ function c4m_message_activity_stream_content_type_render($subtype, $conf, $args,
   $block->module = 'c4m_message';
   $block->title = '';
   $block->content = theme('c4m_message_activity_stream');
+  $block->content .= views_embed_view('group_activity_stream');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/plugins/content_types/activity_stream/activity_stream.inc]: Add a comment

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/filter/filter_homepage.js [hunk 8]
```diff
@@ -11,6 +11,7 @@
       var filter = radios.val();
       window.location.search = '?filter=' + filter;
     });
+
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/filter/filter_homepage.js]: remove this

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 9]
```diff
@@ -3658,3 +3658,83 @@ function c4m_message_get_user_last_activity($uid) {
...
+ *
+ * @param array $og_list
+ *   An array of group IDs to check.
+ * @param string $bundle
+ *   The bundle name to include in the return.
+ *
+ * @return mixed
+ *   An array of the groups of the requested type (bundle).
+ */
+function c4m_message_filter_groups_by_bundle($og_list, $bundle = 'group') {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: remove this function

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 10]
```diff
@@ -3658,3 +3658,83 @@ function c4m_message_get_user_last_activity($uid) {
...
+  if ($view->name != 'global_activity_stream_no_js_fallback') {
+    return;
+  }
+
+  $filterBy = filter_input(INPUT_GET, 'filter', FILTER_SANITIZE_STRING);
+  $groups = og_get_groups_by_user();
+
+  switch ($filterBy) {
+    case 'groups':
+      $groups = c4m_message_filter_groups_by_bundle($groups['node']);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: use `c4m_og_get_user_group_ids_by_bundle()` instead

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 11]
```diff
@@ -3658,3 +3658,83 @@ function c4m_message_get_user_last_activity($uid) {
...
+      $query->where[2]['conditions'][0]['value'] = $groups;
+      $query->where[2]['conditions'][0]['operator'] = 'IN';
+      unset($query->where[2]['conditions'][1]);
+      break;
+
+    case 'projects':
+      $groups = c4m_message_filter_groups_by_bundle($groups['node'], 'project');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: use `c4m_og_get_user_group_ids_by_bundle()` instead

File: project/profiles/capacity4more/themes/c4m/kapablo/theme/preprocess.inc [hunk 12]
```diff
@@ -162,6 +162,13 @@ function kapablo_preprocess_page(&$variables) {
     'label' => t('More info'),
   );
 
+  // Add js to control the body no-js class.
+  drupal_add_js('
+(function ($) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/themes/c4m/kapablo/theme/preprocess.inc]: fix to something like:
```js
jQuery('body').removeClass('no-js');
```

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 13]
```diff
@@ -3658,3 +3658,83 @@ function c4m_message_get_user_last_activity($uid) {
...
+ * Redirect the user to the homepage + query string ?filter=.
+ *
+ * For the global activity stream, when JS is not enabled, this function will
+ * redirect to the correct path when filtering the content by groups, projects
+ * or interests.
+ */
+function c4m_features_homepage_filter_form_create_submit(&$form, &$form_state) {
+  $form_state['redirect'] = array(
+    '<front>',
+    array(
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: move this array to `$options`

File: project/profiles/capacity4more/themes/c4m/kapablo/theme/preprocess.inc [hunk 14]
```diff
@@ -541,6 +541,8 @@ function kapablo_preprocess_image(&$variables) {
 function kapablo_preprocess_html(&$variables) {
   // Render off_canvas into top level variables.
   $variables['off_canvas'] = drupal_render($variables['page']['off_canvas']);
+  // We are adding a class `no-js` to the body element which is set to be
+  // removed by JS. This enables us to easily use fallback css if JS disabled.
```
Comment (ordavidil) [inline project/profiles/capacity4more/themes/c4m/kapablo/theme/preprocess.inc]: not sure how css related in here

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 15]
```diff
@@ -3658,3 +3658,38 @@ function c4m_message_get_user_last_activity($uid) {
     return '';
   }
 }
+
+/**
+ * Implements hook_views_query_alter().
+ */
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: docs

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 16]
```diff
@@ -3658,3 +3658,38 @@ function c4m_message_get_user_last_activity($uid) {
...
+function c4m_message_views_query_alter(&$view, &$query) {
+  if ($view->name != 'global_activity_stream_no_js_fallback') {
+    return;
+  }
+  $filterBy = filter_input(INPUT_GET, 'filter', FILTER_SANITIZE_STRING);
+
+  if ($filterBy == 'groups' || $filterBy == 'projects') {
+    $groups = c4m_og_get_user_group_ids_by_bundle(rtrim($filterBy, 's'));
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: add a comment for the `rtrim` usage

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 17]
```diff
@@ -3658,3 +3658,38 @@ function c4m_message_get_user_last_activity($uid) {
...
+    $query->where[2]['conditions'][0]['value'] = $groups;
+    $query->where[2]['conditions'][0]['operator'] = 'IN';
+  }
+  else {
+    unset($query->where[2]);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: add a comment

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_plugin_argument_default.inc [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+   * Return the node IDs of the user's interests.
+   *
+   * @return string
+   *   A `+` separated list of node IDs of the user's interests.
+   */
+  public function get_argument() {
+    $filterBy = filter_input(INPUT_GET, 'filter', FILTER_SANITIZE_STRING);
+
+    if ($filterBy == 'interests') {
+      global $user;
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_plugin_argument_default.inc]: @ybaras 
please re-roll from develop and use `c4m_user_profile_get_interests()`

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_plugin_argument_default.inc [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+
+// @codingStandardsIgnoreFile
+
+/**
+ * The global activity stream argument default handler.
+ *
+ * To allow filtering the content of the global activity stream by the user's
+ * interests, we need to pass the relevant IDs.
+ */
+class c4m_message_plugin_argument_default extends views_plugin_argument_default {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_plugin_argument_default.inc]: I think we should rename the name of the class to something more meaningful

**Comment by ordavidil** [comment]
@ybaras 
good work, please add a screenshot / gif of the group dashboard as well.

Also, the filters also should be displayed only if relevant, meaning, if the user don't have interests then the _my interests_ filter should not be present (you can check with `president` user), e.g. groups / projects.

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/filter/filter.inc [hunk 20]
```diff
@@ -47,10 +47,11 @@ function c4m_features_homepage_filter_form_create() {
...
 
-  $options += array(
-    'interests' => t('My interests'),
-    'showAll'   => t('Show all'),
-  );
+  if (c4m_user_profile_get_interests()) {
+    $options['interests'] = t('My interests');
+  }
+
+  $options['showAll'] = t('Show all');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/filter/filter.inc]: @ybaras 
good catch, please return early if `$options` is empty, see:
https://github.com/capacity4dev/capacity4more/pull/1466/commits/acbbd2cd33985847d0dc5e42813f70eb0e45f684

**Comment by ybaras** [comment]
**Activity stream by topic -no js**

anim

**Comment by ordavidil** [comment]
@ybaras 
what happens when you try to load more activities ?

**Comment by ybaras** [comment]
**JS enabled:**

screenshot from 2016-11-02 16 25 35

**JS Disabled:**

screenshot from 2016-11-02 16 28 02

**Comment by ybaras** [comment]
**IE, No JS**

screenshot from 2016-11-02 19 23 11

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 21]
```diff
@@ -3657,3 +3657,44 @@ function c4m_message_get_user_last_activity($uid) {
...
+ * Implements hook_views_query_alter().
+ *
+ * Adds the current logged-in user's groups/projects to the filter criteria when
+ * the users chooses to show activity from his groups or projects only.
+ */
+function c4m_message_views_query_alter(&$view, &$query) {
+  if ($view->name != 'global_activity_stream_no_js_fallback') {
+    return;
+  }
+  $filterBy = filter_input(INPUT_GET, 'filter', FILTER_SANITIZE_STRING);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: => `$filter_by`

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 22]
```diff
@@ -3657,3 +3657,44 @@ function c4m_message_get_user_last_activity($uid) {
...
+
+  if ($filterBy == 'groups' || $filterBy == 'projects') {
+    // `rtrim` removes plural 's' from the query string value to get the bundle.
+    $groups = c4m_og_get_user_group_ids_by_bundle(rtrim($filterBy, 's'));
+    $query->where[2]['conditions'][0]['value'] = $groups;
+    $query->where[2]['conditions'][0]['operator'] = 'IN';
+  }
+  else {
+    // Remove the filter criteria when the users chooses to show all activity or
+    // filter by his interests rather than by his groups or projects.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: drop the word _his_

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views.inc [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * @file
+ * Views plugins.
+ */
+
+/**
+ * Implements hook_views_plugins().
+ *
+ * Returns the topics the logged-in user is interested in.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.views.inc]: => Returns the topics (interests) of the current user.

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+
+// @codingStandardsIgnoreFile
+
+/**
+ * The global activity stream argument default handler.
+ *
+ * To allow filtering the content of the global activity stream by the user's
+ * interests, we need to pass the relevant IDs.
+ */
+class c4m_message_user_interests_default extends views_plugin_argument_default {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc]: drop the `default` from the class name

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * @file
+ * Views argument plugin.
+ */
+
+// @codingStandardsIgnoreFile
+
+/**
+ * The global activity stream argument default handler.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc]: not sure about this one ...

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * To allow filtering the content of the global activity stream by the user's
+ * interests, we need to pass the relevant IDs.
+ */
+class c4m_message_user_interests_default extends views_plugin_argument_default {
+
+  /**
+   * Return the node IDs of the user's interests.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc]: interest / topic is a taxonomy term and not a node.

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+   *
+   * @return string
+   *   A `+` separated list of node IDs of the user's interests.
+   */
+  public function get_argument() {
+    $filterBy = filter_input(INPUT_GET, 'filter', FILTER_SANITIZE_STRING);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc]: => `$filter_by`

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+
+    if ($filterBy == 'interests') {
+      // Convert the interest IDs array to a string (1+2+3) for the SQL query.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/includes/c4m_message_user_interests_default.inc]: drop the `for the SQL query`.

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/plugins/content_types/activity_stream/activity_stream.inc [hunk 29]
```diff
@@ -24,6 +24,9 @@ function c4m_message_activity_stream_content_type_render($subtype, $conf, $args,
   $block->module = 'c4m_message';
   $block->title = '';
   $block->content = theme('c4m_message_activity_stream');
+
+  // Add the `group activity stream` view as a fallback if JS is disabled.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/plugins/content_types/activity_stream/activity_stream.inc]: This one is not clear.

File: project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/states/_activity_global.scss [hunk 30]
```diff
@@ -63,3 +63,67 @@
 .message-c4m-insert--node--share {
   margin-top: 25px;
 }
+
+body:not(.no-js) {
+  .field-name-activity-stream-by-topic {
```
Comment (ordavidil) [inline project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/states/_activity_global.scss]: Not sure why do we need all this CSS rules, probably using the `pane-global-activity-stream-no-js-fallback` and `pane-activity-stream` would be enough ...

**Comment by ordavidil** [comment]
- [ ] The filter button should not be displayed when JS is enabled.
- [ ] The filter button doesn't keep all the parameters from the query string (using the pager for example).

**Comment by ybaras** [comment]
Tablet:

anim

Mobile:

anim

**Comment by ordavidil** [comment]
@ybaras 
Please check why the tests fail keep fail, thanks

**Pull request merged**